sympy/functions/combinatorial/factorials.py

```diff
@@ -812,33 +812,10 @@ def _eval(self, n, k):
                     from sympy.core.compatibility import gmpy
                     return Integer(gmpy.bincoef(n, k))
 
-                prime_count_estimate = N(n / log(n))
-
-                # if the number of primes less than n is less than k, use prime sieve method
-                # otherwise it is more memory efficient to compute factorials explicitly
-                if prime_count_estimate < k:
-                    M, result = int(_sqrt(n)), 1
-                    for prime in sieve.primerange(2, n + 1):
-                        if prime > n - k:
-                            result *= prime
-                        elif prime > n // 2:
-                            continue
-                        elif prime > M:
-                            if n % prime < k % prime:
-                                result *= prime
-                        else:
-                            N, K = n, k
-                            exp = a = 0
-
-                            while N > 0:
-                                a = int((N % prime) < (K % prime + a))
-                                N, K = N // prime, K // prime
-                                exp = a + exp
-
-                            if exp > 0:
-                                result *= prime**exp
-                else:
-                    result = ff(n, k) / factorial(k)
+                d = result = n - k + 1
+                for i in range(2, k + 1):
+                    d += 1
+                    result = result // i * d + result % i * d // i
```

Comment:
I this sum preferable to `result * d // i`?

Reply:
Yes, I think it is best to avoid the possibly big multiplication `result * d`, which will be slow.

Reply:
`d` will be relatively small though `result` will grow. I wonder if there would be significant difference in performance.

Reply:
Yes `d` is relatively small as compared to `result` but it is still significant in comparison to `i`.
The former term is a contributing term (no big intermediate result is obtained), but the latter term uses `(i - 1) * d` as the intermediate product which will be significantly smaller than `result * d`.

Reply:
```IPython
In [67]: def f(n, k):
    ...:   d = result = n-k + 1
    ...:   for i in range(2, k + 1):
    ...:     d += 1
    ...:     result = result // i * d + result % i * d // i
    ...:   return result
    ...: 

In [68]: def g(n, k):
    ...:   d = result = n-k + 1
    ...:   for i in range(2, k + 1):
    ...:     d += 1
    ...:     result = result * d // i
    ...:   return result
    ...: 

In [69]: %timeit f(1000, 500)
100 loops, best of 3: 3.09 ms per loop

In [70]: %timeit g(1000, 500)
1000 loops, best of 3: 1.51 ms per loop

In [71]: %timeit f(1000, 50)
10000 loops, best of 3: 153 µs per loop

In [72]: %timeit g(1000, 50)
10000 loops, best of 3: 71.2 µs per loop

In [73]: %timeit f(100, 50)
10000 loops, best of 3: 95.9 µs per loop

In [74]: %timeit g(100, 50)
10000 loops, best of 3: 47.1 µs per loop
```

Reply:
```
In [5]: %timeit f(21232, 42)
1000000 loops, best of 3: 736 ns per loop

In [6]: %timeit g(21232, 42)
1000000 loops, best of 3: 834 ns per loo
```

But for smaller cases, I think the more number of operations (especially modulo and division) cause the slowdown.

Reply:
Interestingly,
```
In [21]: %timeit g(88932, 80)
1000000 loops, best of 3: 756 ns per loop

In [22]: %timeit f(88932, 80)
1000000 loops, best of 3: 1.27 µs per loop
```
For majority cases, `g(n, k)` is faster than `f(n, k)` - tested for around 50 cases.
Should I commit to that?

Thank you @jksuom @isuruf for the review.